FAPI-1419 - Initial Commit

ChrisJBurns opened the pull request:
- Initial commit

Done:
- The bare-bone skeleton of the tool is within this commit
- The tool should allow you to chose a broker address, topic, schema, schema-registry, partition, and offset. It is worth mentioning these are all mandatory
- Currently it is only configured for document-generation-completed, document-generation-failed, document-generation-started and tx-closed, email-send, render-submitted-data-document, filing-received, filing-processed.
- Print deserialised JSON into terminal
- Improve logging to give more information out to user
- Insert help for flags

Resolves [FAPI-1419
](https://companieshouse.atlassian.net/browse/FAPI-1419)

Commit:
Initial Commit

Commit:
Added more topic schemas

Commit:
Add filing-procesed to switch case

marcransome commented on main.go:
```diff
@@ -0,0 +1,161 @@
+            //create new producer
+            p, err := producer.New(&producer.Config{Acks: &producer.WaitForAll, BrokerAddrs: []string{*brokerPtr}})
+            if err != nil {
+                os.Exit(1)
```
Log a suitable error message here to indicate a reason for failure.

ChrisJBurns replied:
Logging improved under this [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/44fff5babb4e1cddf4baa3cbe31a69da8bd39ec2)

marcransome commented on main.go:
```diff
@@ -0,0 +1,161 @@
+
+            // Unmarshal message value and assign it to schemaStruct
+            if err = consumerAvro.Unmarshal(msg.Value, schemaStruct); err != nil {
+                fmt.Println("Error unmarshalling avro:")
```
The `fmt.Println` call here will output a message to the standard output stream rather than the standard error stream, so prefer `fmt.Errorf` and add a format specifier to include the error description as in: `fmt.Errorf("Error unmarshalling avro: %v", err)`.

If the `consumerAvro.Unmarshal()` call fails there would be no need to attempt to republish the message, as `schemaStruct` may contain only empty fields. After outputting the error message, we should treat this as an unrecoverable error and call `os.Exit(1)`.

ChrisJBurns replied:
Logging improved under this [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/44fff5babb4e1cddf4baa3cbe31a69da8bd39ec2)

However, do we want to exit the tool completely if one message fails to unmarshall or do we want to - in this case - `continue` on within the `for` loop?

marcransome commented on main.go:
```diff
@@ -0,0 +1,161 @@
+            // Marshall schemaStruct that contains message value ready for republishing
+            messageBytes, err := producerAvro.Marshal(schemaStruct)
+            if err != nil {
+                fmt.Println("Error Marshalling avro:")
```
As above, output errors to the standard output stream and include the error description: `fmt.Errorf("Error marshalling avro: %v", err)`.

If the `producerAvro.Marshal()` call fails, `messageBytes` will be `nil` and the subsequent message send operation is likely to fail. Call `os.Exit(1)` here after outputting the error message.

ChrisJBurns replied:
Logging improved under this [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/44fff5babb4e1cddf4baa3cbe31a69da8bd39ec2)

However, do we want to exit the tool completely if one message fails to marshall or do we want to - in this case - `continue` on within the `for` loop?

marcransome commented on main.go:
```diff
@@ -0,0 +1,161 @@
+            // republish message
+            partition, offset, err := p.Send(producerMessage)
+            if err != nil {
+                fmt.Println("Error republishing message:")
```
Output errors to the standard output stream and include the error description: `fmt.Errorf("Error republishing message: %v", err)`.

marcransome commented on main.go:
```diff
@@ -0,0 +1,161 @@
+            if err != nil {
+                fmt.Println("Error republishing message:")
+            }
+            fmt.Println("Message republished!", partition, offset)
```
If `partition` and `offset` are part of the message, indicate what they represent: `fmt.Printf("Message republished using partition: %v offset: %v\n", partition, offset)`.

ChrisJBurns replied:
Logging improved under this [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/44fff5babb4e1cddf4baa3cbe31a69da8bd39ec2)

marcransome commented on main.go:
```diff
@@ -0,0 +1,161 @@
+    partitionPtr = flag.Int64("partition", 0, "Partition")
+    zookeeperPtr = flag.String("zookeeper", "", "Zookeeper")
+    offsetPtr = flag.String("offset", "-1", "Offset number")
+    jsonOutPtr = flag.Int("json-out", -1, "Does deserialized JSON get printed to terminal?")
```
The help message descriptions could be more descriptive in some instances (e.g. `Broker address`, `Schema name`) and the description for the `json-out` flag should be reworded as a statement rather than a question, to avoid any ambiguity (e.g. `Print deserialised message JSON`).

ChrisJBurns replied:
Changes [committed](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/7bab91c1429b72f4153905420aafa3dd57607a81)

marcransome commented on main.go:
```diff
@@ -0,0 +1,161 @@
+    fmt.Println("partition:", *partitionPtr)
+    fmt.Println("zookeeper-registry:", *zookeeperPtr)
+    fmt.Println("offset:", *offsetPtr)
+    fmt.Println("json-out:", *jsonOutPtr)
```
It's worth considering the use of `flag.Visit()` or `flag.VisitAll()` here, depending on whether the output should include only the flags that have been set or all flags regardless of value, as in this example:

```
flag.VisitAll(func (f *flag.Flag) {
    fmt.Printf("%v: %v\n", f.Name, f.Value)
})
```

ChrisJBurns replied:
Good point, however if the flags that have only been set are outputted, the user may not know why the tool isn't working as it could be using a defaulted value and because it isn't printed out they'll never know?

marcransome commented on main.go:
```diff
@@ -0,0 +1,161 @@
+)
+
+func main () {
+    flag.Parse()
```
Consider using `flag.PrintDefaults()` at this point to show a usage message describing the flags (and their default values), after testing whether the minimum mandatory flags have been provided or not.

ChrisJBurns replied:
Is it worth putting it as a condition? For example, if the flag `-help` has been passed, to then execute the `PrintDefaults()`? This way, when the user has used the tool correctly, the `PrintDefaults()` isn't printed out as well therefore minimising clutter in the terminal.

marcransome reviewed (commented):
I've added a few comments after looking over the changes briefly, most of which you've already identified as 'todo'. Ideally this should be reviewed by an off-scrum developer once complete.

Reformat your source files to the `golang` standard using `gofmt` and reorder the package imports with `goimports` in a follow-up commit to keep things nicely formatted.

Commit:
Reformat code to adhere to Go standards

Commit:
Organise imports using goimports

ChrisJBurns commented:
Imports and format has been committed

Commit:
Improved description for flag fields

Commit:
Improve error and succes logging

Commit:
Add flag validation and help descriptions for flag

Commit:
Remove commented out print line

Commit:
Format using gofmt and goimports

Commit:
Add separators to make output more readable

Commit:
Add exit statement if error is thrown

Commit:
Improved logging to output progress

Commit:
Added JSON output functionality

Commit:
Remove unused import

Commit:
Use repo schemas instead of local schemas

emead commented on main.go:
```diff
@@ -0,0 +1,225 @@
+			//create new producer
+			p, err := producer.New(&producer.Config{Acks: &producer.WaitForAll, BrokerAddrs: []string{*brokerPtr}})
+			if err != nil {
+				fmt.Errorf("error creating producer: %v", err)
```
I think this won't print the error. Perhaps use `panic` like you have done elsewhere.

ChrisJBurns replied:
Sure, I will replace the error line to just output the message without the error and then add the panic below it. Like so:
`fmt.Println("error creating producer")` 
`panic(err)`
`os.Exit(1)`

Resolved by [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/05095461fe44fbd53f5fbe4aa41f4e1a4be70316)

emead commented on main.go:
```diff
@@ -0,0 +1,225 @@
+	arraySize := make([]int64, 0)
+	if strings.ContainsAny(offset, "-") {
+		slice := strings.Split(offset, "-")
+		minRange, err := strconv.ParseInt(slice[0], 10, 64)
```
`err` not handled

ChrisJBurns replied:
It is handled below that, however thinking about it now... the `err` value could be reset by the  next line so is a good point to check it before as well.

Resolved in [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/97a4f66b580c96b64723b030cb60b2d7ee2cff56)

emead commented on schemas/document-generation-completed.go:
```diff
@@ -0,0 +1,18 @@
+package schemas
+
+//  'document-generation-completed' topic struct
```
Comment should begin with name:
`DocumentGenerationCompleted ...`

ChrisJBurns replied:
Resolved by [0e02a4d](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/0e02a4d3bf784b120dad1a0b2eced82047bb9b16)

emead commented on schemas/document-generation-completed.go:
```diff
@@ -0,0 +1,18 @@
+	DocumentSize          string            `avro:"document_size"        json:"document_size"`
+}
+
+//  'document-generation-completed' topic struct
```
Comment should begin with name:
`DescriptionValues ...`

ChrisJBurns replied:
Resolved by [0e02a4d](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/0e02a4d3bf784b120dad1a0b2eced82047bb9b16)

emead commented on schemas/document-generation-failed.go:
```diff
@@ -0,0 +1,10 @@
+package schemas
+
+//  'document-generation-failed' topic struct
```
Comment should begin with name:
`DocumentGenerationFailed ...`

ChrisJBurns replied:
Resolved by [0e02a4d](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/0e02a4d3bf784b120dad1a0b2eced82047bb9b16)

emead commented on schemas/document-generation-started.go:
```diff
@@ -0,0 +1,7 @@
+package schemas
+
+//  'document-generation-started' topic struct
```
Comment should begin with name:
`DocumentGenerationStarted ...`

ChrisJBurns replied:
Resolved by [0e02a4d](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/0e02a4d3bf784b120dad1a0b2eced82047bb9b16)

emead commented on schemas/email-send.go:
```diff
@@ -0,0 +1,11 @@
+package schemas
+
+//  'email-send' topic struct
```
Comment should begin with name:
`EmailSend ...`

ChrisJBurns replied:
Resolved by [0e02a4d](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/0e02a4d3bf784b120dad1a0b2eced82047bb9b16)

emead commented on schemas/filing-received.go:
```diff
@@ -0,0 +1,35 @@
+package schemas
+
+//  'filing-received' topic struct
```
Comment should begin with name:
`FilingReceived ...`

ChrisJBurns replied:
Resolved by [0e02a4d](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/0e02a4d3bf784b120dad1a0b2eced82047bb9b16)

emead commented on schemas/filing-received.go:
```diff
@@ -0,0 +1,35 @@
+	Items         []Items                  `avro:"items"                           json:"items"`
+}
+
+//  'PresenterFilingRecieved' struct
```
Remove `''` quotes from comment

ChrisJBurns replied:
Resolved by [0e02a4d](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/0e02a4d3bf784b120dad1a0b2eced82047bb9b16)

emead commented on schemas/filing-received.go:
```diff
@@ -0,0 +1,35 @@
+	UserID   string `avro:"user_id"                          json:"user_id"`
+}
+
+//  'SubmissionFilingRecieved' struct
```
Remove `''` quotes from comment

ChrisJBurns replied:
Resolved by [0e02a4d](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/0e02a4d3bf784b120dad1a0b2eced82047bb9b16)

emead commented on schemas/filing-received.go:
```diff
@@ -0,0 +1,35 @@
+	TransactionID string `avro:"transaction_id"                   json:"transaction_id"`
+}
+
+//  'Items' struct
```
Remove `''` quotes from comment

ChrisJBurns replied:
Resolved by [0e02a4d](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/0e02a4d3bf784b120dad1a0b2eced82047bb9b16)

emead commented on schemas/render-submitted-data-document.go:
```diff
@@ -0,0 +1,11 @@
+	ResourceID   string `avro:"resource_id"                          json:"resource_id"`
+	ContentType  string `avro:"content_type"                         json:"content_type"`
+	DocumentType string `avro:"document_type"                        json:"document_type"`
+	UserId       string `avro:"user_id"                              json:"user_id"`
```
Rename to `UserID`

ChrisJBurns replied:
Fixed by [4d074b6](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/4d074b6fa41e47daa856ddb1e1bdfb90acb24692)

emead commented on schemas/render-submitted-data-document.go:
```diff
@@ -0,0 +1,11 @@
+package schemas
+
+//  'render-submitted-data-document' topic struct
```
Comment should begin with name:
`RenderSubmittedDataDocument ...`

ChrisJBurns replied:
Resolved by [0e02a4d](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/0e02a4d3bf784b120dad1a0b2eced82047bb9b16)

emead commented on schemas/schemas.go:
```diff
@@ -0,0 +1,23 @@
+package schemas
+
+//  This method takes the topic name as a parameter
```
Comment should begin with name:
`IdentifySchema takes the ...`

ChrisJBurns replied:
Fixed by [1497ff0](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/1497ff04fe7bf6c7f7b922f303427af079791fde)

emead commented on schemas/tx-closed.go:
```diff
@@ -0,0 +1,7 @@
+package schemas
+
+//  'tx-closed' topic struct
```
Comment should begin with name:
`TxClosed ...`

ChrisJBurns replied:
Resolved by [0e02a4d](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/0e02a4d3bf784b120dad1a0b2eced82047bb9b16)

emead commented on schemas/document-generation-completed.go:
```diff
@@ -0,0 +1,18 @@
+	DescriptionIdentifier string            `avro:"description_identifier"    json:"description_identifier"`
+	DocumentCreatedAt     string            `avro:"document_created_at"       json:"document_created_at"`
+	DescriptionValues     DescriptionValues `avro:"description_values"        json:"description_values"`
+	Location              string            `avro:"location"        json:"location"`
```
Indentation

ChrisJBurns replied:
Fixed by [9806b2e](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/9806b2e8784d0a4aaccea459668699cb96af398b)

emead commented on schemas/document-generation-completed.go:
```diff
@@ -0,0 +1,18 @@
+	DocumentCreatedAt     string            `avro:"document_created_at"       json:"document_created_at"`
+	DescriptionValues     DescriptionValues `avro:"description_values"        json:"description_values"`
+	Location              string            `avro:"location"        json:"location"`
+	DocumentSize          string            `avro:"document_size"        json:"document_size"`
```
Indentation

ChrisJBurns replied:
Fixed by [9806b2e](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/9806b2e8784d0a4aaccea459668699cb96af398b)

emead commented on schemas/filing-received.go:
```diff
@@ -0,0 +1,35 @@
+	ChannelID     string                   `avro:"channel_id"                       json:"channel_id"`
+	Presenter     PresenterFilingRecieved  `avro:"presenter"               json:"presenter"`
+	Submission    SubmissionFilingRecieved `avro:"published_at"                 json:"published_at"`
+	Items         []Items                  `avro:"items"                           json:"items"`
```
Indentation

ChrisJBurns replied:
Fixed by [9806b2e](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/9806b2e8784d0a4aaccea459668699cb96af398b)

Commit:
Add Makefile

Commit:
Fixed waiting loop

Commit:
Added removal line

emead commented on Makefile:
```diff
@@ -0,0 +1,22 @@
+
+.PHONY: build
+build:
+	 rm kafka-management-tool
```
Inconsistent indentation

ChrisJBurns replied:
Fixed by [76377bb](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/76377bb9476dfc57ef4a71077e14f1b377314765)

emead commented on main.go:
```diff
@@ -0,0 +1,231 @@
+			// if so, exit the tool by breaking the ConsumerLoop
+            if consumed == len(offsetArray) {
+                break ConsumerLoop
+            }
```
Please run `gofmt` to correctly format these lines.

Best practise is to configure your editor to automatically format:
https://github.com/companieshouse/styleguides/blob/master/go.md#dev-environments

ChrisJBurns replied:
Good spot, must have missed it on my last commit.

ChrisJBurns replied:
Resolved by [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/05095461fe44fbd53f5fbe4aa41f4e1a4be70316)

emead commented on main.go:
```diff
@@ -0,0 +1,231 @@
+	fmt.Printf("Getting schema for topic: %v \n", *topicPtr)
+	schema, err := schema.Get(*schemaRegistryPtr, *schemaPtr)
+	if err != nil {
+		fmt.Errorf("rror getting schema: %v", err)
```
Spelling: "error"

ChrisJBurns replied:
Good spot.

Resolved by [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/05095461fe44fbd53f5fbe4aa41f4e1a4be70316)

emead commented on main.go:
```diff
@@ -0,0 +1,231 @@
+			fmt.Printf("Republishing message for offset: %v \n", msg.Offset)
+			partition, offset, err := p.Send(producerMessage)
+			if err != nil {
+				fmt.Errorf("error republishing message with offset: %v, %v \n", msg.Offset, err)
```
I think this won't print the error. Perhaps use panic like you have done elsewhere.

ChrisJBurns replied:
Sure, I will replace the error line to just output the message without the error and then add the panic below it. Like so:
`fmt.Printf("Republishing message for offset: %v \n", msg.Offset)` 
`panic(err)`
`os.Exit(1)`

Resolved by [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/05095461fe44fbd53f5fbe4aa41f4e1a4be70316)

emead commented on main.go:
```diff
@@ -0,0 +1,231 @@
+	fmt.Printf("Getting schema for topic: %v \n", *topicPtr)
+	schema, err := schema.Get(*schemaRegistryPtr, *schemaPtr)
+	if err != nil {
+		fmt.Errorf("rror getting schema: %v", err)
```
I think this won't print the error. Perhaps use panic like you have done elsewhere.

ChrisJBurns replied:
Sure, I will replace the error line to just output the message without the error and then add the panic below it. Like so:
`fmt.Println("error getting schema")` 
`panic(err)`
`os.Exit(1)`

Resolved by [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/05095461fe44fbd53f5fbe4aa41f4e1a4be70316)

emead commented on main.go:
```diff
@@ -0,0 +1,231 @@
+	// Unmarshal message value and assign it to schemaStruct
+	fmt.Printf("Unmarshalling message for offset: %v \n", msg.Offset)
+	if err = consumerAvro.Unmarshal(msg.Value, schemaStruct); err != nil {
+		fmt.Errorf("error unmarshalling avro message for offset: %v, %v \n", msg.Offset, err)
```
I think this won't print the error. Perhaps use panic like you have done elsewhere.

ChrisJBurns replied:
Sure, I will replace the error line to just output the message without the error and then add the panic below it. Like so:
`fmt.Errorf("error unmarshalling avro message for offset: %v \n", msg.Offset)` 
`panic(err)`
`os.Exit(1)`

Resolved by [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/05095461fe44fbd53f5fbe4aa41f4e1a4be70316)

emead commented on main.go:
```diff
@@ -0,0 +1,231 @@
+	fmt.Println("Unmarshalled message:")
+	data, err := json.Marshal(schemaStruct)
+	if err != nil {
+		fmt.Errorf("error marshalling JSON message for offset: %v, %v \n", msg.Offset, err)
```
I think this won't print the error. Perhaps use panic like you have done elsewhere.

ChrisJBurns replied:
Sure, I will replace the error line to just output the message without the error and then add the panic below it. Like so:
`fmt.Errorf("error marshalling JSON message for offset: %v \n", msg.Offset)` 
`panic(err)`
`os.Exit(1)`

Resolved by [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/05095461fe44fbd53f5fbe4aa41f4e1a4be70316)

emead commented on main.go:
```diff
@@ -0,0 +1,231 @@
+	fmt.Printf("Marshalling message for offset: %v \n", msg.Offset)
+	message, err := producerAvro.Marshal(schemaStruct)
+	if err != nil {
+		fmt.Errorf("error marshalling avro message for offset: %v, %v \n", msg.Offset, err)
```
I think this won't print the error. Perhaps use panic like you have done elsewhere.

ChrisJBurns replied:
Sure, I will replace the error line to just output the message without the error and then add the panic below it. Like so:
`fmt.Errorf("error marshalling avro message for offset: %v \n", msg.Offset)` 
`panic(err)`
`os.Exit(1)`

Resolved by [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/05095461fe44fbd53f5fbe4aa41f4e1a4be70316)

emead commented on main.go:
```diff
@@ -0,0 +1,231 @@
+			panic(err)
+		}
+
+		index := 0
```
`index` not used - delete

ChrisJBurns replied:
Resolved by [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/6cb5a49fcd24600087b3128739ed504aae6265d3)

emead commented on main.go:
```diff
@@ -0,0 +1,231 @@
+		index := 0
+		for value := minRange; value <= maxRange; value++ {
+			arraySize = append(arraySize, value)
+			index++
```
`index` not used - delete

ChrisJBurns replied:
Resolved by [commit](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/6cb5a49fcd24600087b3128739ed504aae6265d3)

emead commented:
Could do with some unit tests

Commit:
Changes to resolve PR peer review comments

- Removed err printing and replaced with panic(err)
- Formatted to match gofmts standard

Commit:
Add missed error check

Commit:
Removed unused index counter

Commit:
Fix indent inconsistency

Commit:
Correct naming convention

Commit:
Fix indentation inconsistency

ChrisJBurns commented:
Regarding unit tests, it's a good point, however due to it being a tool and not a service related to business I wonder if there is any point in doing tests as they wont be used or needed. Although the argument could be added where they could stop future changes causing regression. Although I think it's more of a nicety at this point.

Commit:
Fix indentation inconsistency

Commit:
Correct comments to go standards

Commit:
Redo comment to adhere to Go standards

Commit:
Fix indentation

Commit:
Merge branch 'develop' into feature/initial-tool-commit

macreman commented:
@cburns93 I'd argue that since this tool is to be used directly against a live service unit tests should be included :) better to add some now as we'll never get around to doing so later

macreman commented on Makefile:
```diff
@@ -0,0 +1,22 @@
+
+.PHONY: build
+build:
+	rm kafka-management-tool
```
should be using standard Makefile targets for this, see https://github.com/companieshouse/styleguides/blob/master/makefiles.md specifically the `clean` target

also, you'll need to add a `-f` flag to this command to avoid `make clean` failing if the binary doesn't exist

ChrisJBurns replied:
The only thing is, the standard make file targets are targetting .zip files and as far as I know .zips can't be made in this situation?

`-f` has been added.

Commited under [087bbdf](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/087bbdfa6d7f6722df3f2759186d53c7050d6f3a)

macreman commented on Makefile:
```diff
@@ -0,0 +1,22 @@
+
+.PHONY: fmt
+fmt:
+	gofmt .
```
`go fmt ./...` will format all go files in the repo

ChrisJBurns replied:
Updated by [60554d5](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/60554d58b1a6cf7037435eca43189d282ad92328)

macreman commented on main.go:
```diff
@@ -0,0 +1,239 @@
+)
+
+// Assigns all flags to variables
+var (
```
any reason we didn't use gofigure for these flags?
https://github.com/companieshouse/gofigure

would make it easier to set up specific jobs allowing the use of environment variables and/or command line flags

ChrisJBurns replied:
I'm not sure if we really need it. Importing an extra library for something that can be done via the normal flags library. Also I don't know when a user would ever use environment variables to run this as its a command line tool

macreman replied:
probably just a taste thing and a nice to have so wouldn't stop the PR going in but for the more environmental options (broker address and schema registry address) it'd be nice to be able to have those sourced in as env vars to make the commands easier to run/remember between environments

ChrisJBurns replied:
Ahh I see, in that case, it indeed would be a nice thing to add. Is it worth creating an issue for it which I can pick up later on? Just so we can get the barebones of the tool into develop so support can start testing it.

macreman commented on main.go:
```diff
@@ -0,0 +1,239 @@
+	schemaPtr         = flag.String("schema", "", "Schema name")
+	schemaRegistryPtr = flag.String("schema-registry", "", "Schema Registry")
+	partitionPtr      = flag.Int64("partition", 0, "Partition (default: 0)")
+	offsetPtr         = flag.String("offset", "", "Offset number")
```
some explanation of the format allowed for this field should be added i.e. the fact that it can be a single integer or a range with `-` used as a separator

ChrisJBurns replied:
Good point, changed under [7efa1c2](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/7efa1c27d45f02043561f0755d11caa68de5f81c)

macreman commented on main.go:
```diff
@@ -0,0 +1,239 @@
+			panic(err)
+		}
+
+		for value := minRange; value <= maxRange; value++ {
```
should probably validate `minRange` <= `maxRange` before creating this array

ChrisJBurns replied:
In what way shall we validate it?

macreman replied:
panic if `minRange > maxRange` as theres probably a typo and we should tell the user

ChrisJBurns replied:
Ahhh I see, I thought you meant if nothing was provided. Good spot, commited under [b9b6873](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/b9b6873ff13253b6e0374d523419d99d2a6dd95e)

macreman commented on main.go:
```diff
@@ -0,0 +1,239 @@
+}
+
+func outputJSON(msg *sarama.ConsumerMessage, messageBytes chan []byte) {
+	schemaStruct := schemas.IdentifySchema(*topicPtr)
```
do we expect every topic we use to have an avro schema assigned to it?

if so then we should be checking for a `nil` value returned here and error.

otherwise the code below should allow for topics without avro schemas and skip the avro schema processing and jump straight to marshalling the json and printing it out

ChrisJBurns replied:
Good spot, forgot about that.. changed under [35d3b3d](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/35d3b3dbcc4359651dbe3fd05564be12fd6b6339)

Commit:
Correct go fmt in make file

Commit:
Correct Makefile

Commit:
Add more description to offset flag

Commit:
Add nill check for schema

Commit:
Add min and max range validation

ChrisJBurns commented:
Am doing the unit tests today, but having some trouble testing `main.go`.. is it something that will stop the PR going in? Or can we add an issue for this which I can then look at later on?

macreman commented:
Support can test on this branch if required, I'd prefer to see some unit tests before merging into develop

Commit:
added unit tests and refactor code

Commit:
unit tests

Commit:
remove commented out code

marcransome commented on main_test.go:
```diff
@@ -0,0 +1,69 @@
+	arrayRange = append(arrayRange, 12)
+	arrayRange = append(arrayRange, 13)
+	arrayRange = append(arrayRange, 14)
+	arrayRange = append(arrayRange, 15)
```
Use slice literals here to reduce the duplication:

```
arraySingle := []int64{10}
arrayRange := []int64{10, 11, 12, 13, 14, 15}
```

ChrisJBurns replied:
Good point, forgot about that.. fixed under [d3782b5](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/d3782b5d24c8067f1a203e68feeb414a57edc41c)

marcransome commented on main_test.go:
```diff
@@ -0,0 +1,69 @@
+	//flag.Set("test.mutexprofile", "test.mutexprofile")
+	//flag.Set("test.outputdir", "test.outputdir")
+	//flag.Set("test.trace", "test.trace")
+	//flag.Set("test.run", "test.run")
```
The commented lines here should be removed if they are not used.

ChrisJBurns replied:
removed

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+
+	"encoding/json"
+
+	"errors"
```
Please run `goimports` against this file (set your editor to do this when saving files)

ChrisJBurns replied:
goimports added to Makefile and imports formatteed under [71f47fa](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/71f47fa67682a0413cde74893d2c9478f897fcba)

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+	// check if all mandatory flags have been provided
+	if err := validateFlags(); err != nil {
+		panic(err)
+		os.Exit(1)
```
No need to call `os.Exit(1)` after `panic`

ChrisJBurns replied:
Removed in [eb8a608](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/eb8a608db6b6d40b8e3056cd2e599a11ebbae986)

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+	if err != nil {
+		fmt.Println("error getting schema")
+		panic(err)
+		os.Exit(1)
```
No need to call `os.Exit(1)` after `panic`

ChrisJBurns replied:
Removed in [eb8a608](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/eb8a608db6b6d40b8e3056cd2e599a11ebbae986)

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+			if err != nil {
+				fmt.Println("error republishing message with offset: %v, \n", msg.Offset)
+				panic(err)
+				os.Exit(1)
```
No need to call `os.Exit(1)` after `panic`

ChrisJBurns replied:
Removed in [eb8a608](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/eb8a608db6b6d40b8e3056cd2e599a11ebbae986)

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+	if err != nil {
+		fmt.Println("error creating producer")
+		panic(err)
+		os.Exit(1)
```
No need to call `os.Exit(1)` after `panic`

ChrisJBurns replied:
Removed in [eb8a608](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/eb8a608db6b6d40b8e3056cd2e599a11ebbae986)

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+	if err = consumerAvro.Unmarshal(msg.Value, schemaStruct); err != nil {
+		fmt.Errorf("error unmarshalling avro message for offset: %v \n", msg.Offset)
+		panic(err)
+		os.Exit(1)
```
No need to call `os.Exit(1)` after `panic`

ChrisJBurns replied:
Removed in [eb8a608](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/eb8a608db6b6d40b8e3056cd2e599a11ebbae986)

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+	if err != nil {
+		fmt.Errorf("error marshalling avro message for offset: %v \n", msg.Offset)
+		panic(err)
+		os.Exit(1)
```
No need to call `os.Exit(1)` after `panic`

ChrisJBurns replied:
Removed in [eb8a608](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/eb8a608db6b6d40b8e3056cd2e599a11ebbae986)

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+	if err != nil {
+		fmt.Errorf("error marshalling JSON message for offset: %v \n", msg.Offset)
+		panic(err)
+		os.Exit(1)
```
No need to call `os.Exit(1)` after `panic`

ChrisJBurns replied:
Removed in [eb8a608](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/eb8a608db6b6d40b8e3056cd2e599a11ebbae986)

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+			fmt.Printf("Republishing message for offset: %v \n", msg.Offset)
+			partition, offset, err := argu.Producer.Send(producerMessage)
+			if err != nil {
+				fmt.Println("error republishing message with offset: %v, \n", msg.Offset)
```
Use `%d` instead of `%v`

ChrisJBurns replied:
fixed under [3cbf6ee](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/3cbf6ee4d21e7253d3130edf3a73bba66e1f0ecd)

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+	fmt.Println("Unmarshalled message:")
+	data, err := json.Marshal(schemaStruct)
+	if err != nil {
+		fmt.Errorf("error marshalling JSON message for offset: %v \n", msg.Offset)
```
This message is not printed (and should not end with a newline)

ChrisJBurns replied:
How come it is not printed?

emead replied:
Errorf formats a message, but does not print it.

https://golang.org/pkg/fmt/#Errorf

ChrisJBurns replied:
commit made under [3afc684](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/3afc6848b4cde1e37833154c4ec383235f9fc8c4) where `Errorf` has been replaced with `Printf`

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+	// Unmarshal message value and assign it to schemaStruct
+	fmt.Printf("Unmarshalling message for offset: %v \n", msg.Offset)
+	if err = consumerAvro.Unmarshal(msg.Value, schemaStruct); err != nil {
+		fmt.Errorf("error unmarshalling avro message for offset: %v \n", msg.Offset)
```
This message is not printed (and should not end with a newline)

ChrisJBurns replied:
How come it is not printed?

emead replied:
Errorf formats a message, but does not print it.

https://golang.org/pkg/fmt/#Errorf

ChrisJBurns replied:
Ah my bad, got confused with PrintF... commit made under [3afc684](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/3afc6848b4cde1e37833154c4ec383235f9fc8c4) where `Errorf` has been replaced with `Printf`

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+	jsonOutPtr        = flag.Int64("json-out", 0, "Print deserialized JSON message (default: 0)")
+)
+
+type Arguments struct {
```
struct should have comment or be unexported

ChrisJBurns replied:
Fixed under [2920384](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/29203845de5e7d2f43fe859e622d96e64b42fea2)

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+	fmt.Printf("Marshalling message for offset: %v \n", msg.Offset)
+	message, err := producerAvro.Marshal(schemaStruct)
+	if err != nil {
+		fmt.Errorf("error marshalling avro message for offset: %v \n", msg.Offset)
```
This message is not printed (and should not end with a newline)

ChrisJBurns replied:
How come it is not printed?

emead replied:
Errorf formats a message, but does not print it.

https://golang.org/pkg/fmt/#Errorf

ChrisJBurns replied:
commit made under [3afc684](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/3afc6848b4cde1e37833154c4ec383235f9fc8c4) where `Errorf` has been replaced with `Printf`

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+		}
+
+		if minRange > maxRange {
+			fmt.Errorf("min range cannot be greater than max range")
```
This message is not printed.

ChrisJBurns replied:
How is it not printed?

emead replied:
Errorf formats a message, but does not print it.

https://golang.org/pkg/fmt/#Errorf

ChrisJBurns replied:
commit made under [3afc684](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/3afc6848b4cde1e37833154c4ec383235f9fc8c4) where `Errorf` has been replaced with `Printf`

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+
+		select {
+		case msg := <-partitionConsumer.Messages():
+			partitionConsumer.Close()
```
error return value not checked.

ChrisJBurns replied:
Which error return value are you referencing?

emead replied:
The error returned by Close()

https://github.com/Shopify/sarama/blob/master/consumer.go#L119

ChrisJBurns replied:
Ahh good spot, err check added under [5462ccc](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/5462ccc5033547c36fc46c0c6d335cdcfdd08360)

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+// related to the offset in the topic/partition is outputted into the 'out' chan
+func consumePartition(consumer sarama.Consumer, topic string, partition int32, out chan *sarama.ConsumerMessage, offsetArray []int64) {
+	for _, offset := range offsetArray {
+		partitionConsumer, err := consumer.ConsumePartition(topic, partition, int64(offset))
```
No need to convert `offset` to `int64`

ChrisJBurns replied:
removed cast in [405fefe](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/405fefe5ee219d86bc4f1cf2f15f6c5294613423)

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+	args := os.Args
+	var error int8 = 0
+	flag.VisitAll(func(f *flag.Flag) {
+		if string(f.Value.String()) == "" && contains(string(f.Name), args) {
```
no need to convert `f.Name` or `f.Value.String()` to strings

ChrisJBurns replied:
remove unnecessary string conversion in [8559af4](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/8559af4dac0d6a64641f4acfb4eff1164ea71c87)

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+			panic(err)
+		}
+
+		select {
```
should use a simple channel send/receive instead of select with a single case

ChrisJBurns replied:
resolved under [e7c40c9](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/e7c40c94b90a6329eb38854a7f126e38fbff43ac)

macreman commented on main.go:
```diff
@@ -179,13 +199,32 @@ func outputJSON(msg *sarama.ConsumerMessage, messageBytes chan []byte) {
+func validateFlags() error {
+	args := os.Args
+	var error int8 = 0
 	flag.VisitAll(func(f *flag.Flag) {
```
This only visits flags that have been passed in on the command line meaning that if you don't pass in a flag it does not get validated.

Test by running the tool with no params, here's the output:

```
$ ./kafka-management-tool
---------------------
Parameters:
broker:
json-out: 0
offset:
partition: 0
schema:
schema-registry:
topic:
---------------------
error creating consumer
panic: kafka: client has run out of available brokers to talk to (Is your cluster reachable?)

goroutine 1 [running]:
main.main()
	main.go:54 +0x379
```

macreman replied:
this seems to be an issue with the custom `contains` function used in the if statement on the line below

macreman replied:
can remove the contains func completely and the reference to it (also remove `args = os.Args` above) as it complicates the issue and is not required

emead commented on main.go:
```diff
@@ -0,0 +1,289 @@
+// throw an error if not the case
+func validateFlags() error {
+	args := os.Args
+	var error int8 = 0
```
`int8` defaults to zero, so you can delete `= 0`

ChrisJBurns replied:
removed in [8b2d3d2](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/8b2d3d2c2caad3b41c7d5c9df939f6e4e9f31100)

emead commented on main_test.go:
```diff
@@ -0,0 +1,59 @@
+	"github.com/Shopify/sarama/mocks"
+	"github.com/companieshouse/chs.go/kafka/producer"
+	. "github.com/smartystreets/goconvey/convey"
+	"testing"
```
Please run `goimports` against this file (set your editor to do this when saving files)

ChrisJBurns replied:
done

emead commented on main_test.go:
```diff
@@ -0,0 +1,59 @@
+		So(createOffsetArray("10"), ShouldResemble, arraySingle)
+		So(createOffsetArray("10-15"), ShouldResemble, arrayRange)
+	})
+}
```
Code coverage can be increased (currently only 40.5%).

[goconvey](http://goconvey.co/) is a handy tool which can be used to view coverage.

marcransome commented on main_test.go:
```diff
@@ -0,0 +1,59 @@
+	arrayRange = append(arrayRange, 14)
+	arrayRange = append(arrayRange, 15)
+
+	Convey("test successful", t, func() {
```
The test description here should state what is being tested.

ChrisJBurns replied:
description add in [9d973a2](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/9d973a26f65842bc2413165487e5c744ede2612b)

emead commented on schemas/document-generation-completed.go:
```diff
@@ -0,0 +1,18 @@
+package schemas
+
+//  DocumentGenerationCompleted struct for document-generation-completed topic
```
Please use only a single space after the double slash.

ChrisJBurns replied:
Remove in [ce005fb](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ce005fb8926cbc2899d61cf66de233a4e80ce97c)

emead commented on schemas/document-generation-completed.go:
```diff
@@ -0,0 +1,18 @@
+	DocumentSize          string            `avro:"document_size"                   json:"document_size"`
+}
+
+//  DescriptionValues struct for document-generation-completed topic
```
Please use only a single space after the double slash.

ChrisJBurns replied:
Remove in [ce005fb](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ce005fb8926cbc2899d61cf66de233a4e80ce97c)

emead replied:
Not done.

ChrisJBurns replied:
double spaces removed in [9edd492](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/9edd49223eaff7f92ad875ac4925a15528511c59)

emead commented on schemas/document-generation-failed.go:
```diff
@@ -0,0 +1,10 @@
+package schemas
+
+//  DocumentGenerationFailed struct for document-generation-failed topic
```
Please use only a single space after the double slash.

ChrisJBurns replied:
Remove in [ce005fb](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ce005fb8926cbc2899d61cf66de233a4e80ce97c)

emead commented on schemas/document-generation-started.go:
```diff
@@ -0,0 +1,7 @@
+package schemas
+
+//  DocumentGenerationStarted struct for document-generation-started topic
```
Please use only a single space after the double slash.

ChrisJBurns replied:
Remove in [ce005fb](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ce005fb8926cbc2899d61cf66de233a4e80ce97c)

emead commented on schemas/email-send.go:
```diff
@@ -0,0 +1,11 @@
+package schemas
+
+//  EmailSend struct for email-send topic
```
Please use only a single space after the double slash.

ChrisJBurns replied:
Remove in [ce005fb](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ce005fb8926cbc2899d61cf66de233a4e80ce97c)

emead commented on schemas/filing-received.go:
```diff
@@ -0,0 +1,35 @@
+package schemas
+
+//  FilingReceived struct for filing-received topic
```
Please use only a single space after the double slash.

ChrisJBurns replied:
Remove in [ce005fb](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ce005fb8926cbc2899d61cf66de233a4e80ce97c)

emead commented on schemas/filing-received.go:
```diff
@@ -0,0 +1,35 @@
+	Items         []Items                  `avro:"items"                           json:"items"`
+}
+
+//  PresenterFilingReceived struct represents PresenterFilingReceived in the FilingReceived struct
```
Please use only a single space after the double slash.

ChrisJBurns replied:
Remove in [ce005fb](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ce005fb8926cbc2899d61cf66de233a4e80ce97c)

emead replied:
Not done.

ChrisJBurns replied:
double spaces removed in [9edd492](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/9edd49223eaff7f92ad875ac4925a15528511c59)

emead commented on schemas/filing-received.go:
```diff
@@ -0,0 +1,35 @@
+	UserID   string `avro:"user_id"                         json:"user_id"`
+}
+
+//  SubmissionFilingReceived struct represents SubmissionFilingReceived in the FilingReceived struct
```
Please use only a single space after the double slash.

ChrisJBurns replied:
Remove in [ce005fb](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ce005fb8926cbc2899d61cf66de233a4e80ce97c)

emead replied:
Not done.

ChrisJBurns replied:
double spaces removed in [9edd492](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/9edd49223eaff7f92ad875ac4925a15528511c59)

emead commented on schemas/filing-received.go:
```diff
@@ -0,0 +1,35 @@
+	TransactionID string `avro:"transaction_id"                  json:"transaction_id"`
+}
+
+//  Items struct represents Items in the FilingReceived struct
```
Please use only a single space after the double slash.

ChrisJBurns replied:
Remove in [ce005fb](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ce005fb8926cbc2899d61cf66de233a4e80ce97c)

emead replied:
Not done.

ChrisJBurns replied:
Do the go imports put double spaces? I committed it with the correct spaces amount but for some reason it has regressed

emead replied:
It shouldn't do - your other spacing changes were committed ok.

ChrisJBurns replied:
I will double check!

ChrisJBurns replied:
double spaces removed in [9edd492](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/9edd49223eaff7f92ad875ac4925a15528511c59)

emead commented on schemas/render-submitted-data-document.go:
```diff
@@ -0,0 +1,11 @@
+package schemas
+
+//  RenderSubmittedDataDocument struct for render-submitted-data-document topic
```
Please use only a single space after the double slash.

ChrisJBurns replied:
Remove in [ce005fb](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ce005fb8926cbc2899d61cf66de233a4e80ce97c)

emead commented on schemas/schemas.go:
```diff
@@ -0,0 +1,23 @@
+package schemas
+
+//  IdentifySchema takes the topic name as a parameter
```
Please use only a single space after the double slash.

ChrisJBurns replied:
Remove in [ce005fb](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ce005fb8926cbc2899d61cf66de233a4e80ce97c)

emead commented on schemas/tx-closed.go:
```diff
@@ -0,0 +1,7 @@
+package schemas
+
+//  TxClosed struct for tx-closed topic
```
Please use only a single space after the double slash.

ChrisJBurns replied:
Remove in [ce005fb](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ce005fb8926cbc2899d61cf66de233a4e80ce97c)

macreman commented on main_test.go:
```diff
@@ -0,0 +1,59 @@
+		consumerGroup.YieldMessage(&sarama.ConsumerMessage{})
+
+		producerMock := mocks.NewSyncProducer(t, nil)
+		producerMock.ExpectSendMessageAndSucceed()
```
This test currently doesn't have any assertions, we should use `ExpectSendMessageWithCheckerFunctionAndSucceed` so that we can ensure the mock producer gets called as expected. For an example of using this you can check the sarama library (mocks folder) https://github.com/Shopify/sarama

ChrisJBurns replied:
commit under [853f08](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/853f08626c590405062a223431bb3e1d9edc2457)

marcransome commented on Makefile:
```diff
@@ -0,0 +1,28 @@
+bin          := kafka-management-tool
```
The `Makefile` doesn't follow the usual conventions we observe for `go` projects and is missing versioning for release candidates, and the generation of `xunit` and `gometalinter` files used by Jenkins. Take a look at https://github.com/companieshouse/styleguides/blob/master/makefiles.md for an example of how we do this in general.

ChrisJBurns replied:
Added under commit [bcc6a1b6](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/bcc6a1b64750e3e39104f192a277cdd51c747514)

Commit:
Add goimports and format imports

Commit:
add description above struct declaration

marcransome commented on main_test.go:
```diff
@@ -0,0 +1,59 @@
+	flag.Set("schema-registry", "schema-registry")
+	flag.Set("topic", "topic")
+
+	Convey("test successful", t, func() {
```
The test description here should state what is being tested.

ChrisJBurns replied:
description add in [9d973a2](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/9d973a26f65842bc2413165487e5c744ede2612b)

Commit:
Remove exit calls as not needed

marcransome commented:
Take a look at the [blog post](https://blog.golang.org/godoc-documenting-go-code) regarding documenting Go code, particularly:

> ...comment is a complete sentence that begins with the name of the element it describes. This important convention allows us to generate documentation in a variety of formats, from plain text to HTML to UNIX man pages, and makes it read better when tools truncate it for brevity, such as when they extract the first line or sentence.

Commit:
replace %v with %d

Commit:
remove assignment due to default

Commit:
remove irrelevant string conversion

Commit:
remove unnecessary cast

Commit:
remove extra space

Commit:
descriptions for unit tests

Commit:
use slice literals instead

Commit:
improved documentation of methods

emead commented on Makefile:
```diff
@@ -6,6 +6,7 @@ all: clean build install
 .PHONY: fmt
 fmt:
 	go fmt ./...
+	goimports ./..
```
I don't think goimports belongs in the makefile (have you tried running it?)
 It's more a tool to configure with your editor.

https://godoc.org/golang.org/x/tools/cmd/goimports

emead commented on Makefile:
```diff
@@ -0,0 +1,29 @@
+.PHONY: fmt
+fmt:
+	go fmt ./...
+	goimports ./..
```
I don't think goimports belongs in the makefile (have you tried running it?)
It's more a tool to configure with your editor.

https://godoc.org/golang.org/x/tools/cmd/goimports

ChrisJBurns replied:
Removed from Makefile and added in my IDE instead

Commit:
replace Errorf with PrintF

Commit:
add error check after channel Close()

Commit:
formatting

Commit:
redo flag validation checking and remove select case

Commit:
add unit tests for extra methods

emead commented on main.go:
```diff
@@ -0,0 +1,304 @@
+	SchemaRegistry string
+	Partition      int64
+	Offset         string
+	JsonOut        int64
```
Change capitalisation to `JSONOut`

ChrisJBurns replied:
done [19892e6](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/19892e64d016e5bf48a8eb8634703e199485f718)

emead commented on main_test.go:
```diff
@@ -0,0 +1,85 @@
+		//value := mocks.ValueChecker(func(value []byte) error {
+		//   lvalue = lvalue + 1
+		//   return nil
+		//})
```
Remove commented code

ChrisJBurns replied:
Good spot

emead commented:
Test coverage can still be increased, e.g. :

- test the printFlags function
- test validateFlags with an invalid map

ChrisJBurns commented:
Not really sure how to test the `printFlags` function if it doesn't return anything?

emead commented:
Fair point. It's not worth testing that function.

Commit:
remove double space in comments

Commit:
add test against invalid map

ChrisJBurns commented:
invalidMaps test done [e3172c0](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/e3172c01d645b0728f9ee3e0708db7f4421bcec0)

Commit:
capitalise json

Commit:
add metalinter xunit targets

Commit:
Amended unit tests

emead commented on schemas/filing-received.go:
```diff
@@ -0,0 +1,35 @@
+	ApplicationID string                   `avro:"application_id"                  json:"application_id"`
+	Attempt       int32                    `avro:"attempt"                         json:"attempt"`
+	ChannelID     string                   `avro:"channel_id"                      json:"channel_id"`
+	Presenter     PresenterFilingRecieved  `avro:"presenter"                       json:"presenter"`
```
Spelling: Received

ChrisJBurns replied:
done under [ac08e34](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ac08e34e0769b621e547e58a3ca6bfa9732febec)

emead commented on schemas/filing-received.go:
```diff
@@ -0,0 +1,35 @@
+	Attempt       int32                    `avro:"attempt"                         json:"attempt"`
+	ChannelID     string                   `avro:"channel_id"                      json:"channel_id"`
+	Presenter     PresenterFilingRecieved  `avro:"presenter"                       json:"presenter"`
+	Submission    SubmissionFilingRecieved `avro:"published_at"                    json:"published_at"`
```
Spelling: Received

ChrisJBurns replied:
done under [ac08e34](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ac08e34e0769b621e547e58a3ca6bfa9732febec)

emead commented on schemas/filing-received.go:
```diff
@@ -0,0 +1,35 @@
+}
+
+// PresenterFilingReceived struct represents PresenterFilingReceived in the FilingReceived struct
+type PresenterFilingRecieved struct {
```
Spelling: Received

ChrisJBurns replied:
done under [ac08e34](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ac08e34e0769b621e547e58a3ca6bfa9732febec)

emead commented on schemas/filing-received.go:
```diff
@@ -0,0 +1,35 @@
+}
+
+// SubmissionFilingReceived struct represents SubmissionFilingReceived in the FilingReceived struct
+type SubmissionFilingRecieved struct {
```
Spelling: Received

ChrisJBurns replied:
done under [ac08e34](https://github.com/companieshouse/kafka-management-tool/pull/1/commits/ac08e34e0769b621e547e58a3ca6bfa9732febec)

Commit:
correct spelling of Received
